lib/lhm/throttler/time.rb

```diff
@@ -5,13 +5,46 @@ class Time
 
       DEFAULT_TIMEOUT = 0.1
       DEFAULT_STRIDE = 2_000
+      DEFAULT_BACKOFF_REDUCTION_FACTOR = 0.2 # 20%
+      MIN_STRIDE_SIZE = 1
 
       attr_accessor :timeout_seconds
       attr_accessor :stride
 
       def initialize(options = {})
         @timeout_seconds = options[:delay] || DEFAULT_TIMEOUT
         @stride = options[:stride] || DEFAULT_STRIDE
+        @backoff_reduction_factor = options[:backoff_reduction_factor] || DEFAULT_BACKOFF_REDUCTION_FACTOR
+        @min_stride_size = options[:min_stride_size] || MIN_STRIDE_SIZE
+
+        if @backoff_reduction_factor >= 1 || @backoff_reduction_factor <= 0
+          raise ArgumentError, 'backoff_reduction_factor must be between greater than 0, and less than 1'
+        end
+
+        if @min_stride_size < 1
+          raise ArgumentError, 'min_stride_size must be and integer greater than 0'
+        end
+
+        if !@min_stride_size.is_a?(Integer)
+          raise ArgumentError, 'min_stride_size must be an integer'
+        end
+
+        if @min_stride_size > @stride
+          raise ArgumentError, 'min_stride_size must be less than or equal to stride'
+        end
+      end
+
+      def backoff_stride
+        new_stride = (@stride * (1 - @backoff_reduction_factor)).to_i
+
+        if new_stride == @stride
+          raise RuntimeError, "Cannot backoff any further"
+        end
```

Comment:
When would this happen? It makes sense looking at the logic, though I'm trying to think of potential cases:

* when `@backoff_reduction_factor` == 0 - This should be caught on initialize.
* when `@stride` == 0 - This should be caught on initialize (`@min_stride_size  >= 1 && @min_stride_size <= @stride`).
* Would rounding be the case here? An additional safeguard?

Curious to know, I may be missing something..

Reply:
You're correct about the rounding! With normal use there shouldn't actually be any situation in which this clause is invoked, it's just meant to be a safeguard against future changes that could potentially introduce an infinite loop. However it is possible to set stride, and backoff_reduction_factor in such a way that results in stride == new_stride (although you really have to be trying to break things)

<img width="572" alt="Screenshot 2024-08-15 at 6 43 50 PM" src="https://github.com/user-attachments/assets/ad153f97-2910-4afc-9e06-f59ea3e87d46">

Reply:
Thanks!